libraries/AP_Networking/AP_Networking.h

```diff
@@ -0,0 +1,196 @@
+
+#pragma once
+
+#include "AP_Networking_Config.h"
+
+#if AP_NETWORKING_ENABLED
+#include <AP_HAL/AP_HAL.h>
+#include "AP_Networking_Params.h"
+
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+    #include "lwipthread.h"
+    #include "lwip/udp.h"
+    #include <lwip/ip_addr.h>
+#else
+    #if BYTE_ORDER == LITTLE_ENDIAN
+    #define IP4_ADDR_VALUE(a,b,c,d)        \
+            (((uint32_t)((d) & 0xff) << 24) | \
+            ((uint32_t)((c) & 0xff) << 16) | \
+            ((uint32_t)((b) & 0xff) << 8)  | \
+            (uint32_t)((a) & 0xff))
+    #else
+    #define IP4_ADDR_VALUE(a,b,c,d)        \
+            (((uint32_t)((a) & 0xff) << 24) | \
+            ((uint32_t)((b) & 0xff) << 16) | \
+            ((uint32_t)((c) & 0xff) << 8)  | \
+            (uint32_t)((d) & 0xff))
+    #endif
+#endif
+
+#define IP4_ADDR_VALUE_FROM_ARRAY(array) IP4_ADDR_VALUE(array[0],array[1],array[2],array[3])
+#define IP4_ADDR_FROM_ARRAY(dest_ip, array) IP4_ADDR(dest_ip, array[0],array[1],array[2],array[3])
+
+#if LWIP_IPV6
+#define IP_ADDR_FROM_ARRAY(dest_ip, array) IP4_ADDR_FROM_ARRAY(dest_ip.u_addr.ip4, array)
+#else
+#define IP_ADDR_FROM_ARRAY(dest_ip, array) IP4_ADDR_FROM_ARRAY(dest_ip, array)
+#endif
+
+// declare backend class
+class AP_Networking_Backend;
+class AP_Networking_SpeedTest;
+class AP_Networking_LatencyTest;
+
+class AP_Networking
+{
+    friend class AP_Networking_Backend;
+    friend class AP_Networking_SpeedTest;
+    friend class AP_Networking_LatencyTest;
+
+public:
+    AP_Networking();
+
+    /* Do not allow copies */
+    CLASS_NO_COPY(AP_Networking);
+
+    struct AP_Networking_State {
+        // TODO: add additional state variables
+        const struct AP_Param::GroupInfo *var_info;
+    };
+    static const struct AP_Param::GroupInfo *backend_var_info[AP_NETWORKING_MAX_INSTANCES];
+
+
+    // options bitmask
+    enum class Options : uint32_t {
+        TFTP                        = (1U<<0),
+        NETBIOS                     = (1U<<1),
+        SNTP                        = (1U<<2),
+    };
+
+    // check if a option is set
+    bool option_is_set(Options option) const {
+        return (uint32_t(_param.options.get()) & uint32_t(option)) != 0;
+    }
+
+    void init();
+
+    void update();
+
+    // Return the number of feature instances
+    uint8_t num_instances(void) const { return _num_instances; }
+    AP_Networking_Params::Type get_type(const uint8_t instance) const;
+
+    static AP_Networking *get_singleton(void) { return _singleton; }
+
+
+    bool is_healthy() const { return _init.done; }    // TODO: is_healthy()
```

Comment:
All the headers returning values like this could add a `WARN_IF_UNUSED`.

Reply:
well... they're pretty much all unused right now so lets not drown going down that path right now... unless you want to chase that.